Do not show WiFi configuration link if there are no WiFi devices

teclator opened the pull request:
## Problem

In #316 we added logic to hide the link for connecting to a WiFi network but only in case the WiFi was enable based on rfkill but if there are no WiFi devices at all then it does not makes sense to show the link.

## Solution

Do not show the link for connecting to a WiFi network if there are no WiFi devices.

Update: and also do not show an empty network section when there is no connection detected and no WiFi device available.

## Testing

- *Added a new unit test*
- *Tested manually*

## Screenshots

| With a network connection | Whitout network connection |
| --- | --- |
| ![Screenshot from 2023-01-02 11-29-04](https://user-images.githubusercontent.com/7056681/210225473-bfaa7277-991a-4b6f-9376-03a8c6d8f913.png) | ![Screenshot from 2023-01-02 11-26-53](https://user-images.githubusercontent.com/7056681/210225333-42842fa9-e4f5-4d40-9b31-d0d18d665632.png) |
| ![Screenshot from 2023-01-02 11-33-33](https://user-images.githubusercontent.com/7056681/210226384-a98afe2a-4594-485d-a69f-40b533edefcb.png) | ![Screenshot from 2023-01-02 11-36-22](https://user-images.githubusercontent.com/7056681/210226383-0d7ddbfe-0527-4f7f-b9dc-36b020f74d9d.png) |

imobachgs commented on web/src/client/network/network_manager.js:
```diff
@@ -500,19 +502,29 @@ class NetworkManagerAdapter {
+  * @return {boolean}
+  */
+  wirelessScanSupported() {
+    const wEnabled = !!(this.proxies.manager?.WirelessEnabled && this.proxies.manager?.WirelessHardwareEnabled);
```
It looks a little hard to read to me. I would rather prefer something like (not tested):

```javascript
if (this.proxies.manager === undefined) return false;

const { WirelessEnabled: wirelessEnabled, WirelessHardwareEnabled: wirelessHardwareEnabled} = this.proxies.manager;
const wifiDevices = Object.values(this.proxies.devices).filter(d => d.DeviceType === NM_DEVICE_TYPE_WIFI);
return WirelessEnabled && wirelessHardware && wifiDevices.length > 0;
```

So return earlier if the manager is unavailable, and make the conditions more readable. Additionally, why do you use "wireless" and "wifi"? It does not look consistent.

teclator replied:
Originally it represented the state of rfill enablement but that is not the case anymore, so, the method and variable should be renamed to WiFi as we are not interested in other kind of devices by now.

teclator replied:
Adapted, I basically renamed the settings property and the method to hopefully a more meaningful name.

imobachgs reviewed (changes requested):
Now that we have plenty of time, I would take the opportunity to improve a few things. For instance, if the network is not initialized, we display [display nothing](https://github.com/yast/d-installer/blob/master/web/src/components/network/Network.test.jsx#L52), and it says nothing to the user. Additionally, if you do not have any network device (e.g., the network cable is disconnected and you do not have any wireless device), you basically see an empty section, right?

teclator commented:
@imobachgs I addressed the suggestions from our previous comment, so, only deciding what to show in case that the section is completely empty.

![Screenshot from 2022-12-16 11-29-26](https://user-images.githubusercontent.com/7056681/208093074-b04ab224-d4c2-4c13-8b81-a5c95911ada0.png)

dgdavid commented:
> @imobachgs I addressed the suggestions from our previous comment, so, only deciding what to show in case that the section is completely empty.
> 

I know we (the YaST team) haven't talk about that yet, but in my opinion the summary mustn't contain an empty section. The section can be there or not, depending on the - not yet ready - configuration (depending on the product? on the architecture? both?), but no empty.

So, I'd add a _summary_ telling the user that D-Installer was not able to find neither, a wired network connection nor a network hardware for connection to a WiFi network.

teclator commented:
> > @imobachgs I addressed the suggestions from our previous comment, so, only deciding what to show in case that the section is completely empty.
> 
> I know we (the YaST team) haven't talk about that yet, but in my opinion the summary mustn't contain an empty section. The section can be there or not, depending on the - not yet ready - [product|architecture|workflow] configuration, but no empty.
> 
> So, I'd add a _summary_ telling the user that D-Installer was not able to find neither, a wired network connection nor a network hardware for connection to a WiFi network.

@dgdavid @imobachgs something like this?

dgdavid commented on web/src/client/network/network_manager.js:
```diff
@@ -500,19 +503,40 @@ class NetworkManagerAdapter {
-      wireless: !!(manager?.WirelessEnabled && manager?.WirelessHardwareEnabled),
-      hostname: settings?.Hostname || ""
-    };
+  /*
+  * Returns the list of WiFi devices available in the system
+  *
+  * @return {object[]} list of available WiFi devices
+  */
+  availableWifiDevices() {
+    return Object.values(this.proxies.devices).filter(d => d.DeviceType === NM_DEVICE_TYPE_WIFI);
+  }
+
+  /*
+  * Returns whether the system is able to scan wifi networks based on rfkill and the presence of
+  * some wifi device
+  *
+  * @return {boolean}
+  */
```
NP: looks to me that those comments are miss-aligned.

```js
/*
*
```

should actually be

```js
/* 
 *
```

dgdavid commented on web/src/client/network/network_manager.js:
```diff
@@ -500,19 +503,40 @@ class NetworkManagerAdapter {
 
   /*
   * Returns NetworkManager general settings
+  *
```
NP: Ah, I see we have a mix on comments style :) So keep them as they are :man_shrugging:

imobachgs replied:
I do not think these are mixed styles, just miss-aligned comments :) And sure we have more of them. 😞

teclator commented:
![Screenshot from 2023-01-02 11-26-53](https://user-images.githubusercontent.com/7056681/210225333-42842fa9-e4f5-4d40-9b31-d0d18d665632.png)

dgdavid reviewed (approved):
LGTM, thank you!

dgdavid reviewed (commented):
Do not forget to update the .changes file.

imobachgs commented on web/package/cockpit-d-installer.changes:
```diff
@@ -1,3 +1,11 @@
+
+- Do not show the link to configure wifi networks when there are no
+  wifi devices and also inform the user when no network connection
+  was detected. (gh#yast/d-installer#323).
```
np: I would remove the first dot

teclator replied:
Ups, fixed :P

imobachgs reviewed (approved):
Just a minor comment in the changes log. Otherwise, LGTM. Thanks!
